packages/app/src/components/inspector/PatternInspector.tsx

```diff
@@ -0,0 +1,238 @@
+const UploadButton = styled.button<{ show: boolean }>(
+  ({ theme, show = false }) => ({
+    color: theme.colors.button.secondaryText,
+    position: 'absolute',
+    background: 'rgba(0, 0, 0, 0.75)',
+    cursor: 'pointer',
+    userSelect: 'auto',
+    alignSelf: 'center',
+    height: '30px',
+    width: '105px',
+    borderRadius: '24px',
+    border: 'none',
+    display: show ? 'block' : 'none',
+    zIndex: 1,
+  }),
+);
+
+interface Props {
+  id: string;
+  images: FileMap;
+  pattern: SketchPattern;
+  onChangeImage?: (image: Sketch.FileRef | Sketch.DataRef) => void;
+  onChangeFillType?: (amount: Sketch.PatternFillType) => void;
+  onChangeTileScale?: (amount: number) => void;
+  createImage: (image: ArrayBuffer, _ref: string) => void;
+}
+
+export type PatternFillTypes = 'Stretch' | 'Fill' | 'Fit' | 'Tile';
+export const patternFillTypeOptions: PatternFillTypes[] = [
+  'Tile',
+  'Fill',
+  'Stretch',
+  'Fit',
+];
+
+export default memo(function PatternInspector({
+  id,
+  images,
+  pattern,
+  onChangeImage,
+  onChangeFillType,
+  onChangeTileScale,
+  createImage,
+}: Props) {
+  const [showButton, setShowButton] = useState(false);
+  const [isDragOver, setIsDragOver] = useState(false);
+
+  const patternType = pattern.patternFillType;
+  const isTile = patternType === Sketch.PatternFillType.Tile;
+
+  const changeFillType = useCallback(
+    (value: PatternFillTypes) => {
+      onChangeFillType?.(Sketch.PatternFillType[value]);
+    },
+    [onChangeFillType],
+  );
+
+  const onSubmitTileScale = useCallback(
+    (value: number) => {
+      onChangeTileScale?.(value / 100);
+    },
+    [onChangeTileScale],
+  );
+
+  const onNudgeTileScale = useCallback(
+    (value: number) => {
+      onChangeTileScale?.(value / 100);
+    },
+    [onChangeTileScale],
+  );
+
+  const background = useMemo(
+    () => getPatternBackground(images, pattern.image),
+    [images, pattern.image],
+  );
+
+  const backgroundSize = useMemo(
+    () => getPatternSize(pattern.patternFillType, pattern.patternTileScale),
+    [pattern.patternFillType, pattern.patternTileScale],
+  );
+
+  const handleImageFile = useCallback(
+    async (file: File) => {
+      const data = await file.arrayBuffer();
+
+      const extension = file.type.split('/')[1];
+      const _ref = `images/${uuid()}.${extension}`;
+
+      createImage(data, _ref);
+      onChangeImage?.({
+        _class: 'MSJSONFileReference',
+        _ref: _ref,
+        _ref_class: 'MSImageData',
+      });
+    },
+    [createImage, onChangeImage],
+  );
+
+  const openFile = useCallback(async () => {
+    const file = await fileOpen({
+      extensions: imgFileExtensions.map((e) => `.${e}`),
+      mimeTypes: mimeTypes,
+    });
+
+    handleImageFile(file);
+  }, [handleImageFile]);
+
+  const handleDragEvent = useCallback((e: DragEvent, on?: boolean) => {
+    if (on) setIsDragOver(on);
+    e.preventDefault();
+    // Do something
+  }, []);
+
+  const handleDropEvent = useCallback(
```

Comment:
Let's try separating the drag and drop handling into a `hooks/useFileDropTarget.ts`.

```ts
function useFileDropTarget() {
  // useState stuff
  return {
    dropTargetProps: {
      onDragOver: ...,
      onDragEnter: ...,
      onDragLeave: ...,
      onDrop: ...
    },
    isDropTargetActive, // maybe instead of `isDropOver`
  }
}
```

Then this component can use it:

```tsx
const { dropTargetProps, isDropTargetActive } = useFileDropTarget()

return (
  // ...
    <Container {...dropTargetProps} isActive={isDropTargetActive}>...</Container>
  // ...
)
```